core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -58,27 +63,53 @@ public AtomicFileWriter(File f) throws IOException {
     }
 
     /**
-     * @param encoding
-     *      File encoding to write. If null, platform default encoding is chosen.
+     * @param encoding File encoding to write. If null, platform default encoding is chosen.
+     *
+     * @deprecated Use {@link #AtomicFileWriter(Path, Charset)}
      */
-    public AtomicFileWriter(File f, String encoding) throws IOException {
-        File dir = f.getParentFile();
-        try {
-            dir.mkdirs();
-            tmpFile = File.createTempFile("atomic",null, dir);
-        } catch (IOException e) {
-            throw new IOException("Failed to create a temporary file in "+ dir,e);
-        }
-        destFile = f;
-        if (encoding==null)
-            encoding = Charset.defaultCharset().name();
+    @Deprecated
+    public AtomicFileWriter(@Nonnull File f, @Nullable String encoding) throws IOException {
+        this(toPath(f), encoding == null ? Charset.defaultCharset() : Charset.forName(encoding));
+    }
+
+    /**
+     * Wraps potential {@link java.nio.file.InvalidPathException} thrown by {@link File#toPath()} in an
+     * {@link IOException} for backward compatibility.
+     *
+     * @param file
+     * @return the path for that file
+     * @see File#toPath()
+     */
+    private static Path toPath(@Nonnull File file) throws IOException {
         try {
-            core = new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding));
+            return file.toPath();
         } catch (InvalidPathException e) {
             throw new IOException(e);
         }
     }
 
+    /**
+     * @param destinationPath the destination path where to write the content when committed.
+     * @param charset File charset to write.
+     */
+    public AtomicFileWriter(@Nonnull Path destinationPath, @Nonnull Charset charset) throws IOException {
+        if (charset == null) { // be extra-defensive if people don't care
+            throw new IllegalArgumentException("charset is null");
+        }
+        this.destPath = destinationPath;
+        Path dir = this.destPath.getParent();
+        try {
+            if (Files.notExists(dir)) { // This is required for dir symlink handling, see JDK-8130464
```

Comment:
should this be more defensive and check that if the file does exist that it is either a symlink to a directory or a directory?

```java
if (Files.exists(dir) && ((!isDirectory(dir)))  {
 throw new IOException(dir + " exists and is neither a direcotry nor a symlink to a directory");
}
else {
  Files.createDirectories(dir);
}
```

Reply:
+1